### ISSUE TITLE
Indexer Setup for Chain State Data Retrieval

### ISSUE THREAD

**Issue by annatekl** [pendulum-chain/spacewalk#473]
This ticket is for indexer setup in order to  Enable efficient retrieval of aggregated data over multiple blocks for Spacewalk.

### Requirements: 
1. Ensure the indexer is configured to efficiently handle data retrieval from various blocks. 

2. Retrieval of current chain state data for Spacewalk: 
- Amount issued per asset
- Number of external vaults (if any)
- Total collateral locked by external vaults (if any)
- Total number of issue and redeem transactions

**Comment by annatekl** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Pendulum-6408a0365cda59c57d39bee5/issues/pendulum-chain/spacewalk/473?planning-poker) with Zenhub @b-yap @bogdanS98 @ebma @gianfra-t @TorstenStueber

**Comment by gianfra-t** [comment]
For some of these points I think is better to check the chain state directly like we were discussing in [the slack](https://satoshipay.slack.com/archives/C01V1F56RMJ/p1704394545532149?thread_ts=1704380893.065049&cid=C01V1F56RMJ). 
That doesn't mean it cannot be done in the indexer though, we can keep track of it also, it is just that the most reliable source for any current "snapshot" will always be the chain itself.

**Comment by TorstenStueber** [comment]
Yes, I also think that we should always directly use the chain state if this already gives sufficient information so that we can keep our indexers as simple as possible.

Such snapshot information is:
* Amount issued per asset
* Number of external vaults (the chain does not know what an external vault is but we can retrieve all vaults and then exclude our own ones by a script)

Possibly also the total collateral locked, however, there is some kind of aggregation script over all vaults required.

**Comment by ebma** [comment]
> Ensure the indexer is configured to efficiently handle data retrieval from various blocks.

I don't understand this requirement. That's a basic feature of every block indexer, no?

---
For the rest

> Amount issued per asset

Can be queried with "Chain State" > `tokens::totalIssuance(currency)`

> Number of external vaults (if any)

As @TorstenStueber mentioned, internal vs external vaults is something that we make up and it doesn't make sense to implement this distinction into the indexer. 

> Total collateral locked by external vaults (if any)

Can be queried for all vaults with "Chain State" > `vaultStaking::totalStake(0, None)`

> Total number of issue and redeem transactions

All issue and redeem requests can be queried with "Chain State" > `issue::issueRequests(None)` or `redeem::redeemRequests(None)`


From just the chain state it would be very difficult to tell these things for a specific timeframe though. So if you wanted to know 'How many XLM were issued from 01.12.23 - today?" we'd need to touch the indexer after all.

**Comment by gianfra-t** [comment]
The problem with total issuance from the state is that it will not be possible to see the accumulated right? Only the snapshot.

**Comment by gianfra-t** [comment]
Yes I was referring to that. So from the chain we get like you said the issue amount subtracted the redeemed. But perhaps another useful metric (to be defined of course) is the total accumulated bridged amount over time. And for that we need something like the squids I guess.

**Comment by ebma** [comment]
@annatekl can you move this ticket to the pendulum-squids repository instead?
